tests/e2e/main.go

```diff
@@ -0,0 +1,433 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		},
+	}
+
+	log.Infof("Creating %q namespace...", nsObj.Name)
+	if _, err := testConfig.KubeClient.CoreV1().Namespaces().Get(nsObj.Name, metav1.GetOptions{}); err != nil {
+		if _, err := testConfig.KubeClient.CoreV1().Namespaces().Create(nsObj); err != nil {
+			return fmt.Errorf("unable to create namespace: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func createCRD(testConfig *TestConfig, crd *apiextensionsv1beta1.CustomResourceDefinition) error {
+	log.Infof("Creating %q CRD...", crd.Name)
+	crcClient := testConfig.APIExtensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions()
+	if _, err := crcClient.Get(crd.Name, metav1.GetOptions{}); err != nil {
+		if _, err := crcClient.Create(crd); err != nil {
+			return fmt.Errorf("unable to create CRD: %v", err)
+		}
+	}
+
+	return wait.Poll(pollInterval, timeoutPoolInterval, func() (bool, error) {
+		if _, err := crcClient.Get(crd.Name, metav1.GetOptions{}); err == nil {
+			return true, nil
+		}
+
+		log.Infof("Waiting for %q crd to be created", crd.Name)
+		return false, nil
+	})
+}
+
+func createConfigMap(testConfig *TestConfig, configMap *apiv1.ConfigMap) error {
+	log.Infof("Creating %q ConfigMap...", strings.Join([]string{configMap.Namespace, configMap.Name}, "/"))
+	if _, err := testConfig.KubeClient.CoreV1().ConfigMaps(configMap.Namespace).Get(configMap.Name, metav1.GetOptions{}); err != nil {
+		if _, err := testConfig.KubeClient.CoreV1().ConfigMaps(configMap.Namespace).Create(configMap); err != nil {
+			return fmt.Errorf("unable to create ConfigMap: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func createSecret(testConfig *TestConfig, secret *apiv1.Secret) error {
+	log.Infof("Creating %q secret...", strings.Join([]string{secret.Namespace, secret.Name}, "/"))
+	if _, err := testConfig.KubeClient.CoreV1().Secrets(secret.Namespace).Get(secret.Name, metav1.GetOptions{}); err != nil {
+		if _, err := testConfig.KubeClient.CoreV1().Secrets(secret.Namespace).Create(secret); err != nil {
+			return fmt.Errorf("unable to create secret: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func createDeployment(testConfig *TestConfig, deployment *appsv1beta2.Deployment) error {
+	log.Info("Creating machine-api-operator...")
+	deploymentsClient := testConfig.KubeClient.AppsV1beta2().Deployments(deployment.Namespace)
+	if _, err := deploymentsClient.Get(deployment.Name, metav1.GetOptions{}); err != nil {
+		if _, err := deploymentsClient.Create(deployment); err != nil {
+			return fmt.Errorf("unable to create Deployment: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// binary runner
+func cmdRun(assetsDir string, binaryPath string, args ...string) error {
+	cmd := exec.Command(binaryPath, args...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Dir = assetsDir
+	return cmd.Run()
+}
+
+var rootCmd = &cobra.Command{
+	Use:   "e2e",
+	Short: "Test deployment of cluster-api stack",
+	RunE: func(cmd *cobra.Command, args []string) error {
```

Comment:
we should break this into couple more functions like `prepareTestEnv()` and `runTest()`. May be in a follow-up PR.